core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java

```diff
@@ -735,7 +735,14 @@ private void failStuckSplitTasks()
                             log.warn("%s is long running with stackTrace:\n%s", splitInfo.getSplitInfo(), stackTraceElements.stream().map(Object::toString).collect(joining(lineSeparator())));
                         }
 
-                        return stuckSplitStackTracePredicate.test(stackTraceElements);
+                        boolean isThreadStack = stuckSplitStackTracePredicate.test(stackTraceElements);
+
+                        // We check if thread's name matches threadId stored in RunningSplitInfo to be sure that stacktrace we obtained belongs to the execution of split
+                        // described by RunningSplitInfo.
+                        // There is still a chance that we may observe the stacktrace from execution of new split before thread name is set in io.trino.execution.executor.TaskExecutor.TaskRunner.run()
+                        // Yet, we assume that such stacktrace would not be classified as "stack" by stuckSplitStackTracePredicate.
+                        boolean splitAssignmentDidNotChange = splitInfo.getThread().getName().startsWith(splitInfo.getThreadId());
```

Comment:
There are other possible ways to do this. Some random ideas:
* Create our own version of Thread that contains additional information that we can inspect
* Drive the process of finding stuck threads from the list of all running splits instead of trying to infer the split by inspecting the threads
* Maintain an external mapping of split <-> thread

Also, for any approach that relies on looking at the thread and then checking if it corresponds to the split we want to kill, there will always be a small window for a race condition unless we force proper synchronization of these checks with split-to-thread assignments.

Reply:
Yeah. I agree this is somewhat brittle. For my defence - there is test which (at least partially) verifies it works ;).
The simple change to that, keeping the same solution principle, would be to have thread <-> split-execution-id mapping, where split-execution-id would be incremented any time split assigned to a thread changes.

And yeah - there is a race-condition window still - I mention that in the comment.